haystack/utils/augment_squad.py

```diff
@@ -0,0 +1,178 @@
+from tqdm import tqdm
+
+logger = logging.getLogger(__name__)
+
+def load_glove(glove_path: Path = Path("glove.txt"), vocab_size: int = 100_000):
+    if not glove_path.exists():
+        zip_path = glove_path.parent / (glove_path.name + ".zip")
+        request = requests.get("https://nlp.stanford.edu/data/glove.42B.300d.zip", allow_redirects=True)
+        with zip_path.open("wb") as f:
+            f.write(request.content)
+        with ZipFile(zip_path, "r") as zip_file:
+            glove_file = zip_file.namelist()[0]
+            with glove_path.open("wb") as g:
+                g.write(zip_file.read(glove_file))
+
+    word_id_mapping = {}
+    id_word_mapping = {}
+    vectors = []
+    with open(glove_path, "r") as f:
+        for i, line in enumerate(f):
+            if i == vocab_size:
+                break
+            split = line.split()
+            word_id_mapping[split[0]] = i
+            id_word_mapping[i] = split[0]
+            vectors.append(np.array([float(x) for x in split[1:]]))
+    vectors = np.stack(vectors)
+    vectors = vectors / np.linalg.norm(vectors, axis=1)[:, np.newaxis]
+    return word_id_mapping, id_word_mapping, vectors
+
+def tokenize_and_extract_words(text, tokenizer):
+    words = tokenizer.basic_tokenizer.tokenize(text)
+
+    subwords = [tokenizer.wordpiece_tokenizer.tokenize(word) for word in words]
+
+    word_subword_mapping = {}
+
+    j = 0
+    for i, subwords_ in enumerate(subwords):
+        if len(subwords_) == 1:
+                word_subword_mapping[i] = j + 1
+        j += len(subwords_)
+    
+    subwords = [subword for subwords_ in subwords for subword in subwords_] # flatten list of lists
+
+    input_ids = tokenizer.convert_tokens_to_ids(subwords)
+    input_ids.insert(0, tokenizer.cls_token_id)
+    input_ids.append(tokenizer.sep_token_id)
+
+    return input_ids, words, word_subword_mapping
+
+def get_replacements(glove_word_id_mapping, glove_id_word_mapping, glove_vectors, model: BertForMaskedLM, tokenizer: BertTokenizer, text, word_possibilities=20):
+    input_ids, words, word_subword_mapping = tokenize_and_extract_words(text, tokenizer)
+
+    batch = []
+    for word_index in word_subword_mapping:
+        subword_index = word_subword_mapping[word_index]
+        input_ids_ = copy(input_ids)
+        input_ids_[subword_index] = tokenizer.mask_token_id
+        batch.append(input_ids_)
+    
+    if batch:
+        batch = torch.tensor(batch)
+        predictions = model(input_ids=batch)
+
+    possible_words = []
+
+    batch_index = 0
+    for i, word in enumerate(words):
+        if i in word_subword_mapping:
+            subword_index = word_subword_mapping[i]
+            logits = predictions["logits"][batch_index, subword_index]
+            ranking = torch.argsort(logits, descending=True)[:word_possibilities]
+            possible_words.append([word] + tokenizer.convert_ids_to_tokens(ranking))
+
+            batch_index += 1
+        elif word in glove_word_id_mapping:
```

Comment:
this is where we could try fasttext. The elif will become an else then because there are no out-of-vocabulary issues with fasttext 
```
import fasttext
ft = fasttext.load_model('cc.en.300.bin') 
#german model: https://dl.fbaipublicfiles.com/fasttext/vectors-crawl/cc.de.300.bin.gz
ft.get_nearest_neighbors('hello')
```

Reply:
That could be useful as 43% of the times when BERT can't be used glove doesn't work either. (Same test on [this dataset](https://github.com/deepset-ai/haystack/blob/master/test/samples/squad/tiny.json) as above)

However, I'm not so sure how great the difference would really be as this would still only be the case in 2.6% of all cases at the cost of being comparable to the original paper or at the cost of additional distillation runs (as explained above they take up a lot of time).